tests/e2e/steps_inactive_vals.go

```diff
@@ -0,0 +1,554 @@
+								ValidatorID("alice"): false,
+								ValidatorID("bob"):   false,
+								ValidatorID("carol"): true,
+							},
+						},
+					},
+				},
+			},
+
+			// unjail alice
+			{
+				Action: UnjailValidatorAction{
+					Provider:  ChainID("provi"),
+					Validator: ValidatorID("alice"),
+				},
+				State: State{
+					ChainID("provi"): ChainState{
+						ValPowers: &map[ValidatorID]uint{
+							ValidatorID("alice"): 100, // alice is back as an active consensus validator
+							ValidatorID("bob"):   0,   // bob is still jailed
+							ValidatorID("carol"): 500,
+						},
+					},
+				},
+			},
+			// unjail bob
+			{
+				Action: UnjailValidatorAction{
+					Provider:  ChainID("provi"),
+					Validator: ValidatorID("bob"),
+				},
+				State: State{
+					ChainID("provi"): ChainState{
+						ValPowers: &map[ValidatorID]uint{
+							ValidatorID("alice"): 0,   // alice is back out because only 2 validators can be active in consensus
+							ValidatorID("bob"):   196, // bob is back as an active consensus validator and lost 2 more power due to the second downtime
+							ValidatorID("carol"): 500,
+						},
+					},
+				},
+			},
+			// relay the packets so that the consumer chain is in sync again
+			{
+				Action: RelayPacketsAction{
+					ChainA:  ChainID("provi"),
+					ChainB:  ChainID("consu"),
+					Port:    "provider",
+					Channel: 0,
+				},
+				State: State{
+					ChainID("consu"): ChainState{
+						ValPowers: &map[ValidatorID]uint{
+							ValidatorID("alice"): 100, // both alice and bob are validating the consumer
+							ValidatorID("bob"):   196,
+							ValidatorID("carol"): 500,
+						},
+					},
+				},
+			},
+		},
+	)
+
+	return s
+}
+
+// Precondition: The provider chain is running.
+// Postcondition: A consumer chain with Top N = 0 is running, including an up-and-running IBC connection to the provider.
+// "alice", "bob", "carol" have opted in and are validating.
+func setupOptInChain() []Step {
+	return []Step{
+		{
+			Action: SubmitConsumerAdditionProposalAction{
+				Chain:         ChainID("provi"),
+				From:          ValidatorID("alice"),
+				Deposit:       10000001,
+				ConsumerChain: ChainID("consu"),
+				SpawnTime:     0,
+				InitialHeight: clienttypes.Height{RevisionNumber: 0, RevisionHeight: 1},
+				TopN:          0,
+			},
+			State: State{
+				ChainID("provi"): ChainState{
+					Proposals: &map[uint]Proposal{
+						1: ConsumerAdditionProposal{
+							Deposit:       10000001,
+							Chain:         ChainID("consu"),
+							SpawnTime:     0,
+							InitialHeight: clienttypes.Height{RevisionNumber: 0, RevisionHeight: 1},
+							Status:        "PROPOSAL_STATUS_VOTING_PERIOD",
+						},
+					},
+					HasToValidate: &map[ValidatorID][]ChainID{
+						ValidatorID("alice"): {},
+						ValidatorID("bob"):   {},
+						ValidatorID("carol"): {},
+					},
+				},
+			},
+		},
+		// Οpt in "alice" and "bob" so the chain is not empty when it is about to start. Note, that "alice" and "bob" use
```

Comment:
This is a setup function, so why not opt in all 3 validators at once? What's the benefit of opting in `bob` after the chain starts?

Reply:
None, just copy-pasted this. Will remove, thx for pointing out :)